sympy/combinatorics/rewritingsystem.py

```diff
@@ -359,93 +359,74 @@ def _add_to_automaton(self, rules):
+            # Transitions will be modified only when suffixes of the current_state
+            # belongs to the proper_prefixes of the new rules.
             # The rest are ignored if they cannot lead to a dead state after a finite number of transisitons.
             if current_state_type == 's':
                 for letter in automaton_alphabet:
                     if letter in self.reduction_automaton.states:
                         self.reduction_automaton.states[state].add_transition(letter, letter)
                     else:
                         self.reduction_automaton.states[state].add_transition(letter, current_state_name)
-            else:
-                # Store subwords of current state in an array.
-                current_state_name_array = []
+            elif current_state_type == 'a':
                 next_sub = current_state_name
-                while len(next_sub):
-                    current_state_name_array.append(next_sub)
-                    next_sub = next_sub.subword(1, len(next_sub))
                 # Check if the transition to any new state in posible.
-                for sub_word in current_state_name_array:
-                    if sub_word in accept_states:
+                while len(next_sub):
+                    if next_sub in accept_states:
                         # Add accept states.
-                        if current_state_type == 'a':
-                            for letter in automaton_alphabet:
-                                next = current_state_name*letter
-                                while True:
-                                    if len(next) <= 1:
-                                        if next in self.reduction_automaton.states:
-                                            self.reduction_automaton.states[state].add_transition(letter, next)
-                                        else:
-                                            self.reduction_automaton.states[state].add_transition(letter, 'start')
-                                        break
-                                    else:
-                                        if next in self.reduction_automaton.states:
-                                            self.reduction_automaton.states[state].add_transition(letter, next)
-                                            break
-                                        next = next.subword(1, len(next))
-                        break
+                        for letter in automaton_alphabet:
+                            _next = current_state_name*letter
+                            while len(_next):
+                                if _next in self.reduction_automaton.states:
+                                    self.reduction_automaton.states[state].add_transition(letter, _next)
+                                    break
+                                elif len(_next) == 1:
```

Comment:
Should this not be `len(_next) == 0`?
Also, if you were to do
```
while len(_next) and _next not in self.reduction_automaton.states:
    _next = _next.subword(1, len(_next))
if not len(_next):
    _next = 'start'
self.reduction_automaton.states[state].add_transition(letter, _next)
```
that would be a bit shorter.

Reply:
Alright 👍